compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java

```diff
@@ -212,29 +222,335 @@ public void visit(BLangPackage pkgNode) {
+            return false;
+        }
+        return false;
+    }
+
+    private void createClosureMapUpdateExpression(BLangClassDefinition classDef,
+                                                  BVarSymbol blockMap,
+                                                  BVarSymbol classMapSymbol) {
+
+        OCEDynamicEnvironmentData oceEnvData = classDef.oceEnvData;
+        BVarSymbol oceMap = oceEnvData.mapBlockMapSymbol;
+
+        BLangFunction function = classDef.generatedInitFunction;
+
+        // self
+        BVarSymbol selfSymbol = function.receiver.symbol;
+        BLangSimpleVarRef selfVarRef = ASTBuilderUtil.createVariableRef(function.pos, selfSymbol);
+        // $map$block$_2
+        BLangSimpleVarRef refToBlockClosureMap = ASTBuilderUtil.createVariableRef(function.pos, blockMap);
+
+        // self.$map$objectCtor
+        BLangIdentifier identifierNode = ASTBuilderUtil.createIdentifier(function.pos, blockMap.name.value);
+        BLangFieldBasedAccess fieldAccess = ASTBuilderUtil.createFieldAccessExpr(selfVarRef, identifierNode);
+        fieldAccess.symbol = oceMap;
+        fieldAccess.setBType(classMapSymbol.type);
+        fieldAccess.expectedType = classMapSymbol.type;
+        fieldAccess.isStoreOnCreation = true;
+        fieldAccess.isLValue = true;
+        fieldAccess.fieldKind = FieldKind.SINGLE;
+        fieldAccess.leafNode = true;
+
+        // self.$map$objectCtor = $map$block$_2
+        BLangAssignment assignmentStmt = (BLangAssignment) TreeBuilder.createAssignmentNode();
+        assignmentStmt.expr = refToBlockClosureMap;
+        assignmentStmt.pos = function.pos;
+        assignmentStmt.setVariable(fieldAccess);
+        fieldAccess.parent = assignmentStmt;
+
+        BLangFunction generatedInitFunction = classDef.generatedInitFunction;
+        BLangBlockFunctionBody generatedInitFnBody = (BLangBlockFunctionBody) generatedInitFunction.body;
+
+        // self[$map$objectCtor$_<num1>] = $map$block$_<num2>
+        generatedInitFnBody.stmts.add(0, assignmentStmt);
+        assignmentStmt.parent = generatedInitFnBody;
+
+        desugar.rewrite(assignmentStmt, oceEnvData.objMethodsEnv);
+    }
+
+    // TODO: filter out only the needed variables, otherwise OCE has indrect access to all block level
+    private void addClosureMapToInit(BLangClassDefinition classDef, BVarSymbol mapSymbol) {
+
+        OCEDynamicEnvironmentData oceData = classDef.oceEnvData;
+
+        // eg : $map$block$_<num2>
+        BLangSimpleVarRef refToBlockClosureMap = ASTBuilderUtil.createVariableRef(classDef.pos, mapSymbol);
+        BLangTypeInit typeInit = oceData.typeInit;
+        BLangInvocation initInvocation = typeInit.initInvocation;
+        if (typeInit.argsExpr == null) {
+            typeInit.argsExpr = new ArrayList<>();
+        }
+
+        // update new expression
+        typeInit.argsExpr.add(refToBlockClosureMap);
+        initInvocation.requiredArgs.add(refToBlockClosureMap);
```

Comment:
We can remove this also https://github.com/ballerina-platform/ballerina-lang/issues/34395